src/libstrongswan/plugins/openssl/openssl_kem.c

```diff
@@ -0,0 +1,371 @@
+}
+
+METHOD(key_exchange_t, get_public_key, bool, private_key_exchange_t *this,
+	chunk_t *value)
+{
+	/* responder action */
+	if (this->ciphertext.ptr != NULL)
+	{
+		*value = chunk_clone(this->ciphertext);
+		return TRUE;
+	}
+
+	/* initiator action */
+	if (!this->pkey && !openssl_kem_generate_pkey(this))
+	{
+		return FALSE;
+	}
+	return openssl_kem_get_encoded_public_key(this, value);
+}
+
+METHOD(key_exchange_t, get_shared_secret, bool, private_key_exchange_t *this,
+	chunk_t *secret)
+{
+	if (this->shared_secret.ptr != NULL)
+	{
+		*secret = chunk_clone(this->shared_secret);
+		return TRUE;
+	}
+	return FALSE;
+}
+
+/**
+ * Decrypt a ciphertext using the private key from the context and store a
+ * reference to the resulting shared secret.
+ */
+static bool openssl_kem_decapsulate(private_key_exchange_t *this,
+									chunk_t ciphertext)
+{
+	EVP_PKEY_CTX *ctx;
+	size_t shared_secret_length = 0;
+
+	ctx = EVP_PKEY_CTX_new(this->pkey, NULL);
+	if (!ctx)
+	{
+		return FALSE;
+	}
+	if (!EVP_PKEY_decapsulate(ctx, NULL, &shared_secret_length, ciphertext.ptr,
+							  ciphertext.len))
+	{
+		EVP_PKEY_CTX_free(ctx);
+		return FALSE;
+	}
+	this->shared_secret = chunk_alloc(shared_secret_length);
+	if (!EVP_PKEY_decapsulate(ctx, this->shared_secret.ptr,
+							  &this->shared_secret.len, ciphertext.ptr,
+							  ciphertext.len))
+	{
+		EVP_PKEY_CTX_free(ctx);
+		chunk_clear(&this->shared_secret);
+		return FALSE;
+	}
+	EVP_PKEY_CTX_free(ctx);
+	return TRUE;
+}
+
+/**
+ * Generate and store a reference to a shared secret. Use the provided public
+ * key to encrypt the shared secret and store a reference to the ciphertext.
+ */
+static bool openssl_kem_encapsulate(private_key_exchange_t *this,
+									chunk_t public_key)
+{
+	EVP_PKEY_CTX *ctx;
+	EVP_PKEY *pkey;
+	size_t shared_secret_length = 0;
+	size_t ciphertext_length = 0;
+	size_t seed_length = 0;
+	chunk_t seed = chunk_empty;
+
+	pkey = EVP_PKEY_kem_new_raw_public_key(openssl_kem_get_nid(this),
+										   public_key.ptr, public_key.len);
+	if (!pkey)
+	{
+		return FALSE;
+	}
+	ctx = EVP_PKEY_CTX_new(pkey, NULL);
+	if (!ctx)
+	{
+		EVP_PKEY_free(pkey);
+		return FALSE;
+	}
+	if (this->drbg)
+	{
+		if (!EVP_PKEY_encapsulate_deterministic(ctx, NULL, &ciphertext_length,
+												NULL, &shared_secret_length,
+												NULL, &seed_length))
+		{
+			EVP_PKEY_free(pkey);
+			EVP_PKEY_CTX_free(ctx);
+			return FALSE;
+		}
+		this->shared_secret = chunk_alloc(shared_secret_length);
+		this->ciphertext = chunk_alloc(ciphertext_length);
+		seed = chunk_alloc(seed_length);
+
+		if (!this->drbg->generate(this->drbg, seed.len, seed.ptr) ||
+			!EVP_PKEY_encapsulate_deterministic(ctx, this->ciphertext.ptr,
+												&this->ciphertext.len,
+												this->shared_secret.ptr,
+												&this->shared_secret.len,
+												seed.ptr, &seed.len))
+		{
+			EVP_PKEY_free(pkey);
+			EVP_PKEY_CTX_free(ctx);
+			chunk_clear(&this->shared_secret);
+			chunk_free(&this->ciphertext);
```

Comment:
You don't have to free/clear these explicitly as that's happening in `destroy()` anyway. (However, unlike with `this->pkey` there is no problem with double-frees,  because `chunk_clear/free()` sets the passed pointer to `chunk_empty`.)